File: android/src/main/java/expo/modules/furiganatext/ReactNativeFuriganaTextModule.kt
package expo.modules.furiganatext

import android.graphics.Color
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

private fun parseColorString(color: String): Int {
  return try {
    Color.parseColor(color)
  } catch (e: IllegalArgumentException) {
    val hex = if (color.startsWith("#")) color.substring(1) else color
    when (hex.length) {
      3 -> {
        val r = hex[0].toString().repeat(2)
        val g = hex[1].toString().repeat(2)
        val b = hex[2].toString().repeat(2)
        Color.parseColor("#$r$g$b")
      }
      4 -> {
        val r = hex[0].toString().repeat(2)
        val g = hex[1].toString().repeat(2)
        val b = hex[2].toString().repeat(2)
        val a = hex[3].toString().repeat(2)
        Color.parseColor("#$a$r$g$b")
      }
      else -> Color.BLACK
    }
  }
}

class ReactNativeFuriganaTextModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("ReactNativeFuriganaText")

    Function("measureHeight") { text: String, furigana: Map<String, String>, width: Double, fontSize: Double, furiganaFontSize: Double ->
      val density = appContext.reactContext?.resources?.displayMetrics?.density ?: 1f
      val (textPaint, furiganaPaint) =
        FuriganaTextLayout.paints(fontSize.toFloat(), furiganaFontSize.toFloat(), density)
      val layout = FuriganaTextLayout.build(
        text, furigana, textPaint, furiganaPaint,
        (width * density).toInt()
      )
      layout.height / density.toDouble()
    }

    View(ReactNativeFuriganaTextView::class) {
      Prop("text") { view: ReactNativeFuriganaTextView, text: String ->
        view.setText(text)
      }

      Prop("furigana") { view: ReactNativeFuriganaTextView, furigana: Map<String, String> ->
        view.setFurigana(furigana)
      }

      Prop("fontSize") { view: ReactNativeFuriganaTextView, fontSize: Double ->
        view.setTextSize(fontSize.toFloat())
      }

      Prop("color") { view: ReactNativeFuriganaTextView, color: String ->
        view.setTextColor(parseColorString(color))
      }

      Prop("furiganaFontSize") { view: ReactNativeFuriganaTextView, fontSize: Double ->
        view.setFuriganaTextSize(fontSize.toFloat())
      }

      Prop("furiganaColor") { view: ReactNativeFuriganaTextView, color: String ->
        view.setFuriganaTextColor(parseColorString(color))
      }

      Prop("fontFamily") { view: ReactNativeFuriganaTextView, fontFamily: String? ->
        view.setFontFamily(fontFamily)
      }

      Prop("fontWeight") { view: ReactNativeFuriganaTextView, fontWeight: String? ->
        view.setFontWeight(fontWeight)
      }

      Prop("fontStyleValue") { view: ReactNativeFuriganaTextView, fontStyle: String? ->
        view.setFontStyleValue(fontStyle)
      }
    }
  }
}

File: ios/ReactNativeFuriganaTextModule.swift
import ExpoModulesCore
import CoreText
import UIKit

extension UIColor {
  convenience init?(hex: String) {
    var hexString = hex
    if hexString.hasPrefix("#") {
      hexString.removeFirst()
    }
    var rgba: UInt64 = 0
    Scanner(string: hexString).scanHexInt64(&rgba)
    let r, g, b, a: CGFloat
    switch hexString.count {
    case 8:
      r = CGFloat((rgba >> 24) & 0xFF) / 255.0
      g = CGFloat((rgba >> 16) & 0xFF) / 255.0
      b = CGFloat((rgba >> 8) & 0xFF) / 255.0
      a = CGFloat(rgba & 0xFF) / 255.0
    case 6:
      r = CGFloat((rgba >> 16) & 0xFF) / 255.0
      g = CGFloat((rgba >> 8) & 0xFF) / 255.0
      b = CGFloat(rgba & 0xFF) / 255.0
      a = 1.0
    default:
      return nil
    }
    self.init(red: r, green: g, blue: b, alpha: a)
  }
}

public class ReactNativeFuriganaText: Module {
  public func definition() -> ModuleDefinition {
    Name("ReactNativeFuriganaText")

    View(ReactNativeFuriganaTextView.self) {
      Prop("text") { (view: ReactNativeFuriganaTextView, text: String) in
        view.text = text
      }

      Prop("furigana") { (view: ReactNativeFuriganaTextView, furigana: [String: String]) in
        view.furigana = furigana
      }

      Prop("fontSize") { (view: ReactNativeFuriganaTextView, fontSize: Double) in
        view.fontSize = CGFloat(fontSize)
      }

      Prop("color") { (view: ReactNativeFuriganaTextView, color: String) in
        view.textColor = UIColor(hex: color) ?? .black
      }

      Prop("furiganaFontSize") { (view: ReactNativeFuriganaTextView, fontSize: Double) in
        view.furiganaFontSize = CGFloat(fontSize)
      }

      Prop("furiganaColor") { (view: ReactNativeFuriganaTextView, color: String) in
        view.furiganaColor = UIColor(hex: color) ?? .gray
      }

      Prop("fontFamily") { (view: ReactNativeFuriganaTextView, fontFamily: String) in
        view.fontFamily = fontFamily
      }

      Prop("fontWeight") { (view: ReactNativeFuriganaTextView, fontWeight: String) in
        view.fontWeight = fontWeight
      }

      Prop("fontStyleValue") { (view: ReactNativeFuriganaTextView, fontStyle: String) in
        view.fontStyleValue = fontStyle
      }
    }
  }
}